---
title: "AVITI_03_mapping"
output: html_document
date: "2023-08-20"
---

# Align trimmed reads to the reference genome HISAT2
## 1. Installation 
```{bash}
git clone https://github.com/DaehwanKimLab/hisat2.git
cd hisat2
make
# Version 2.2.1
export PATH=/home/exouser/hisat2/:$PATH
```

## 2. Building Genome indexes
### 2.1 First, extract splice sites and exons from the Araport11 GTF file. 
```{bash}
cd /media/volume/sde/reference_genome
~/hisat2/hisat2_extract_splice_sites.py Araport11_GTF_genes_transposons.current.gtf > Araport11.ss
~/hisat2/hisat2_extract_exons.py Araport11_GTF_genes_transposons.current.gtf > Araport11.exon
```

### 2.2 Then, build a HISAT2 index using the transcriptome annotation files built above.
```{bash}
~/hisat2/hisat2-build \
-p 4 \
--exon ../Araport11.exon \
--ss ../Araport11.ss \
../TAIR10_chr_all.fas \
Index
```

## 3 Alignment
```{bash}
cd /media/volume/sde/aligned_reads_HISAT2/

# create a file named 'names.txt' with library letter codes (e.g. 'B1') in each row
for i in $(cat /media/volume/sde/file_names/names.txt)
do \
~/hisat2/hisat2 \
-p 4 \
-q \
--min-intronlen 20 \
--max-intronlen 12000 \
-x /media/volume/sde/reference_genome/HISAT2_indices/Index \
-1 /media/volume/sde/trimmed_reads/trimmed_${i}_forward_paired.fq.gz \
-2 /media/volume/sde/trimmed_reads/trimmed_${i}_reverse_paired.fq.gz \
-S HISAT2_${i}_Aligned.sam \
--novel-splicesite-outfile HISAT2_${i}_novel_ss.tab \
--summary-file HISAT2_${i}_summary.txt \
--new-summary

samtools view -bS HISAT2_${i}_Aligned.sam > HISAT2_${i}_Aligned.bam
samtools sort HISAT2_${i}_Aligned.bam -o HISAT2_${i}_Aligned.sorted.bam

rm HISAT2_${i}_Aligned.sam 
rm HISAT2_${i}_Aligned.bam

done
```

This creates 3 files: 
1. alignment .sam file 'HISAT2_*_Aligned.sam'; 
2. HISAT2 reports a list of splice sites in the file 'HISAT2_*_novel_ss.tab'; 
3. alignment summary 'HISAT2_*_summary.txt'
